Fix cycle error with existential types

Aaron1011 opened the pull request:
Fixes #61863

We now allow uses of `existential type`'s that aren't defining uses - that is, uses which don't constrain the underlying concrete type.

To make this work correctly, we also modify `eq_opaque_type_and_type` to not try to apply additional constraints to an opaque type. If we have code like this:

```rust
existential type Foo;
fn foo1() -> Foo { ... }
fn foo2() -> Foo { foo1() }
```

then `foo2` doesn't end up constraining `Foo`, which means that `foo2` will end up using the type `Foo` internally - that is, an actual `TyKind::Opaque`. We don't want to equate this to the underlying concrete type - we just need to enforce the basic equality constraint between the two types (here, the return type of `foo1` and the return type of `foo2`)

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
r? @oli-obk

Centril commented:
cc @rust-lang/lang

Centril commented on src/test/run-pass/existential_type_const.rs:
```diff
@@ -0,0 +1,17 @@
+#![feature(existential_type)]
+#![feature(impl_trait_in_bindings)]
```
Why is `impl_trait_in_bindings` used here?

Aaron1011 replied:
`#![feature(existential_type)]` implicitly depends on `#![feature(impl_trait_in_bindings)]` to work properly - specifically, [this code](https://github.com/rust-lang/rust/blob/481068a707679257e2a738b40987246e0420e787/src/librustc_typeck/check/mod.rs#L856) needs to run, so that we properly visit all defining uses of an existential type. Should I add a comment explaining that?

Centril replied:
A comment would be helpful indeed. :)

Centril commented on src/test/run-pass/existential_type_const.rs:
```diff
@@ -0,0 +1,17 @@
+#![feature(impl_trait_in_bindings)]
+//~^ WARN the feature `impl_trait_in_bindings` is incomplete and may cause the compiler to crash
+
+// Ensures that consts can constrain an existential type
```
```suggestion
// Ensures that `const` items can constrain an `existential type`.
```

Centril commented on src/test/run-pass/existential_type_const.rs:
```diff
@@ -0,0 +1,17 @@
+#![feature(existential_type)]
```
Please move this to `ui/existential_types/` and use a `// check-pass` annotation since there are no dynamic semantics under test. The same should apply to the other new test as well.

Centril commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1253,17 +1253,38 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         .at(&ObligationCause::dummy(), param_env)
                         .eq(output_ty, revealed_ty)?);
 
+                    // This is 'true' when we're using an existential
```
```suggestion
                    // This is `true` when we're using an existential
```

Centril commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1253,17 +1253,38 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         .eq(output_ty, revealed_ty)?);
 
+                    // This is 'true' when we're using an existential
+                    // type without 'revelaing' it. For example, code like this:
```
```suggestion
                    // type without "revealing" it. For example, code like this:
```

Centril commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1253,17 +1253,38 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
+                    // fn foo1() -> Foo { ... }
+                    // fn foo2() -> Foo { foo1() }
+                    //
+                    // In 'foo2', we're not revealing the type of 'Foo' - we're
```
```suggestion
                    // In `foo2`, we're not revealing the type of `Foo` - we're
```

Centril commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1253,17 +1253,38 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
+                    //
+                    // In 'foo2', we're not revealing the type of 'Foo' - we're
+                    // just treating it as the opaque type. All of the constraints
+                    // in our 'opaque_type_map' apply to the concrete type,
```
```suggestion
                    // in our `opaque_type_map` apply to the concrete type,
```

Aaron1011 commented:
My initial solution didn't work properly when an existential type was used inside another type (e.g. `fn foo() -> (Foo, bool)` where `Foo` is an existential type). I've modified my fix to handle that properly, and added an additional test case.

Centril commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1253,17 +1253,38 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
+                    // just treating it as the opaque type. All of the constraints
+                    // in our 'opaque_type_map' apply to the concrete type,
+                    // not to the opaque type itself. Therefore, it's enough
+                    // to simply equate the output and opque 'revealed_type',
```
```suggestion
                    // to simply equate the output and opaque `revealed_type`,
```

Centril commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1253,17 +1253,38 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
+                    // in our 'opaque_type_map' apply to the concrete type,
+                    // not to the opaque type itself. Therefore, it's enough
+                    // to simply equate the output and opque 'revealed_type',
+                    // as we do above
```
```suggestion
                    // as we do above.
```

Centril commented on src/librustc_typeck/check/writeback.rs:
```diff
@@ -576,36 +576,38 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
-                            opaque_defn,
-                            old,
-                        );
+                if !skip_add {
```
Could use comments here?

Centril commented:
Since you added a commit mid my review they are sorta outdated now but the basic idea isn't so you can "translate" the changes to the new location... ;)

Aaron1011 commented:
@Centril: I've addressed your comments

cramertj commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1268,15 +1268,43 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         let opaque_defn_ty = tcx.type_of(opaque_def_id);
                         let opaque_defn_ty = opaque_defn_ty.subst(tcx, opaque_decl.substs);
                         let opaque_defn_ty = renumber::renumber_regions(infcx, &opaque_defn_ty);
+                        let concrete_is_opaque = infcx
```
I'm not sure if this is quite the right check, or at least it doesn't match the behavior specified by the RFC. In the RFC, it's expected that the following would compile:

```rust
existential type Foo: Debug;

fn foo() -> Foo {
    5u32
}

fn bar(cond: bool) -> Foo {
    if cond { foo() } else { 5u32 }
}
```

In this situation, the type of the variable returned by `foo` is an impl trait variable, but since the function is within the same module as the definition of the `existential type`, it can constrain the type further and itself become a defining use. The RFC says you can even do this:

```rust
existential type Foo: Debug;

fn foo() -> Foo {
    5u32
}

fn bar() -> Foo {
    let x: u32 = foo();
    x + 5
}
```

I realize this isn't at all what's implemented today, but it'd be good to talk about how the behavior being implemented here interacts with the desired behavior we want to end up with.

Aaron1011 replied:
I believe that this check is still compatible with implementing the full behavior specified in the RFC. This check occurs after we've attempted to instantiate the opaque type. If we're still left with an opaque type after instantiation, then there's nothing else we can do.

Once the rest of the RFC is implemented, this check should be hit only when the opaque type is being used outside the defining module.

cramertj replied:
That makes sense-- thanks for the clarification!

eddyb commented:
r? @nikomatsakis

Aaron1011 commented:
@nikomatsakis: We always check that the opaque type is equal to the 'output' type [here](https://github.com/rust-lang/rust/blob/8ba9b1019c6e6c514826c5466e84d93f665f975f/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L1267-L1269). We only skip applying the obligations that should only apply to the underlying revealed type.

As I mentioned to @cramertj, I believe this change is fully forwards-compatible with 'revealing' opaque types under more circumstances (e.g. within the defining module). My checks runs during MIR typecheck - at this stage of compilation, any opauqe tyoe usages that have not been revelaed are never going to be reveled (i.e. usages outside the defining module). Future changes may affect under what circumstances my check is hit, but I believe that it's always correct to skip the 'revealed' obligations for an opaque type that we cannot reveal.

Aaron1011 commented:
@nikomatsakis: Are there any changes that you'd like me to make?

oli-obk commented:
r? @oli-obk

oli-obk commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1268,15 +1268,43 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
+                        // without actually looking at the underlying type it
+                        // gets 'revealed' into
+
+                        if !concrete_is_opaque {
```
if the concrete type is opaque, don't we need an additional check to figure out if the opaque types are equal? Essentially what I'm asking is whether

```rust
pub trait MyTrait {}

#[derive(Debug)]
pub struct MyStruct {
  v: u64
}

impl MyTrait for MyStruct {}

mod foo {
    pub fn bla() -> TE {
        return crate::MyStruct {v:1}
    }
    
    
    existential type TE: crate::MyTrait;
}

mod bar {
    pub fn bla2() -> TE2 {
        crate::foo::bla()
    }
    
    pub existential type TE2: crate::MyTrait;
}

pub fn bla2() -> bar::TE2 {
    crate::foo::bla()
}
```

still errors after your PR

Aaron1011 replied:
@oli-obk See my reply to @nikomatsakis above

oli-obk replied:
The code you linked is to the master branch (or at least before your PR) right here. So we aren't always checking that as you wrote.

Aaron1011 replied:
Yes - that line is [unchanged]9https://github.com/Aaron1011/rust/blob/2f4196205336df8550a4bfb3045d3d1c350f02bf/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L1263-L1265) in my PR - that is, we still equate the opaque types.

oli-obk replied:
:man_facepalming: 	sorry, These two obligations looked so similar, I confused the two.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 2f4196205336df8550a4bfb3045d3d1c350f02bf has been approved by `oli-obk`

<!-- @bors r=oli-obk 2f4196205336df8550a4bfb3045d3d1c350f02bf -->
<!-- homu: {"type":"Approved","sha":"2f4196205336df8550a4bfb3045d3d1c350f02bf","approver":"oli-obk"} -->
